napari/layers/base/base.py

```diff
@@ -678,6 +678,10 @@ def metadata(self, value: dict) -> None:
     def source(self) -> Source:
         return self._source
 
+    @source.setter
+    def source(self, source: Source) -> None:
+        self._source = source
```

Comment:
Oops, I just saw now you wrote about this in the description @DragaDoncila ^^' consider this my answer!

Reply:
@brisvag thanks for the review. I couldn't think of a way to update the context manager to update the source, because the context manager just sets a "singleton" source that layers query on init - whereas in this case the layer is already built. I looked at whether we could just move the `with layer_source` call up to where we ask the plugin to build the layer, but we need part of the return value from the function to get the plugin name for the source (for npe1 compatibility purposes 🥲 ), so that would've required considerable surgery. I will address this latter issue at some stage, but in the meantime I've made a private source setter that doesn't allow overwriting the source. Does that seem reasonable? Happy to hear other ideas!

Reply:
@DragaDoncila Maybe just use context manager when call plugin, so if user creates a layer in it, it will set source using context manager information?

Reply:
@Czaki as mentioned in the comment above, we use some of the information returned from the plugin call to set the source, for npe1 compatibility reasons, so this would require some additional surgery.

Reply:
> in the meantime I've made a private source setter that doesn't allow overwriting the source. Does that seem reasonable? Happy to hear other ideas!

that seems like a reasonable compromise to me

Reply:
Still, may we do this for npe2 plugins? 
We may do not even support returning layers from npe1 plugins.

Reply:
@Czaki well, we will support returning layers from npe1 plugins, after this PR merges, because the code that actually reads files is "unified" in a single function that gets called from `add_layers_with_plugins`. This function resolves the npe1/npe2 reader, reads the data, returns the `Layer` or `LayerDataTuple` **and** returns the name of the plugin that actually did the reading, which is what we set as source. The change in this PR will therefore actually also work for npe1 plugins.

I don't think this reading cascade is "a good thing"™️ but I also don't think it should be changed in this PR, just for the sake of not adding a private function that, realistically, nobody is going to see or care about except us, when we set the source of an already constructed layer.